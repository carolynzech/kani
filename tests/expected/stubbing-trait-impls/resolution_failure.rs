// Copyright Kani Contributors
// SPDX-License-Identifier: Apache-2.0 OR MIT
// kani-flags: -Z stubbing
//! Check that Kani supports stubbing methods in trait implementations
//! using the motivating example from https://github.com/model-checking/kani/issues/1997

trait A {
    fn foo(&self) -> u32;

    fn bar(&self) -> u32;
}

trait B {
    fn bar(&self) -> u32;
}

struct X {}

impl X {
    fn new() -> Self {
        Self {}
    }
}

impl A for X {
    fn foo(&self) -> u32 {
        100
    }

    fn bar(&self) -> u32 {
        200
    }
}

impl B for X {
    fn bar(&self) -> u32 {
        300
    }
}

#[kani::proof]
// A::baz doesn't exist
#[kani::stub(<X as B>::bar, A::baz)]
// X::bar is ambiguous; two impls could match
#[kani::stub(X::bar, A::foo)]
// B::bar doesn't have a body
#[kani::stub(X::foo, B::bar)]
fn resolution_errors() {}
