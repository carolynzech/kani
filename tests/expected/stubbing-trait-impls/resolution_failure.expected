error: invalid stub; function does not have a body, but is not an extern function\
resolution_failure\
|\
| #[kani::stub(X::foo, B::bar)]\
| ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
|\
note: B::bar` does not have a body\
resolution_failure\
|\
|     fn bar(&self) -> u32;\
|     ^^^^^^^^^^^^^^^^^^^^^\
   = help: if the stub refers to associated functions, try using fully-qualified syntax instead

error: failed to resolve `X::bar`: unable to find `bar` inside struct `X`\
resolution_failure\
|\
| #[kani::stub(X::bar, A::foo)]\
| ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\
|

error: failed to resolve `A::baz`: unable to find `baz` inside trait `A`\
resolution_failure\
|\
| #[kani::stub(<X as B>::bar, A::baz)]\
| ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\
|

error: aborting due to 3 previous errors; 4 warnings emitted
